crates/prover/src/components/store_deref_fp.rs

```diff
@@ -0,0 +1,468 @@
+// -(fp+off2, [prev_value, 0, 0, 0], prev_clock) || +(fp+off2, [value, 0, 0, 0], clock)
+// -(fp+off0, [value, 0, 0, 0], prev_clock) || +(fp+off0, [value, 0, 0, 0], clock)
+const N_MEMORY_LOOKUPS: usize = 2 * 3;
+const N_REGISTERS_LOOKUPS: usize = 2; // -(pc, fp) || +(pc+1, fp)
+
+const LOOKUPS_COLUMNS: usize = N_MEMORY_LOOKUPS + N_REGISTERS_LOOKUPS;
+
+pub struct InteractionClaimData {
+    pub lookup_data: LookupData,
+    pub non_padded_length: usize,
+}
+
+#[derive(Uninitialized, IterMut, ParIterMut)]
+pub struct LookupData {
+    // 6 elements: addr, value_0, value_1, value_2, value_3, clock
+    pub memory: [Vec<[PackedM31; 6]>; N_MEMORY_LOOKUPS],
+    // 2 elements: pc, fp
+    pub registers: [Vec<[PackedM31; 2]>; N_REGISTERS_LOOKUPS],
+}
+
+/// The enabler column is a column of length `padding_offset.next_power_of_two()` where
+/// 1. The first `padding_offset` elements are set to 1;
+/// 2. The rest are set to 0.
+#[derive(Debug, Clone)]
+pub struct Enabler {
+    pub padding_offset: usize,
+}
+impl Enabler {
+    pub const fn new(padding_offset: usize) -> Self {
+        Self { padding_offset }
+    }
+
+    pub fn packed_at(&self, vec_row: usize) -> PackedM31 {
+        let row_offset = vec_row * N_LANES;
+        if self.padding_offset <= row_offset {
+            return PackedM31::zero();
+        }
+        if self.padding_offset >= row_offset + N_LANES {
+            return PackedM31::one();
+        }
+
+        // The row is partially enabled.
+        let mut res = [M31::zero(); N_LANES];
+        for v in res.iter_mut().take(self.padding_offset - row_offset) {
+            *v = M31::one();
+        }
+        PackedM31::from_array(res)
+    }
+}
+
+#[derive(Clone, Default, Serialize, Deserialize)]
+pub struct Claim {
+    pub log_size: u32,
+}
+
+impl Claim {
+    pub fn mix_into(&self, channel: &mut impl Channel) {
+        channel.mix_u64(self.log_size as u64);
+    }
+
+    pub fn log_sizes(&self) -> TreeVec<Vec<u32>> {
+        let trace = vec![self.log_size; N_TRACE_COLUMNS];
+        // TODO: check the correct width of vector for the interaction trace
+        let interaction_trace = vec![self.log_size; SECURE_EXTENSION_DEGREE * LOOKUPS_COLUMNS];
+        TreeVec::new(vec![vec![], trace, interaction_trace])
+    }
+
+    pub fn write_trace<MC: MerkleChannel>(
+        inputs: &mut Vec<StateData>,
+    ) -> (Self, ComponentTrace<N_TRACE_COLUMNS>, InteractionClaimData)
+    where
+        SimdBackend: BackendForChannel<MC>,
+    {
+        let non_padded_length = inputs.len();
+        let log_size = std::cmp::max(LOG_N_LANES, inputs.len().next_power_of_two().ilog2());
+
+        let (mut trace, mut lookup_data) = unsafe {
+            (
+                ComponentTrace::<N_TRACE_COLUMNS>::uninitialized(log_size),
+                LookupData::uninitialized(log_size - LOG_N_LANES),
+            )
+        };
+        inputs.resize(1 << log_size, StateData::default());
```

Comment:
this mutation of the input is still a question mark for me,

1/ for performance, we don't want to clone the data,
2/ for cleanliness, we probably shouldnt mutate an input